worker/codalabworker/local_run/local_run_state.py

```diff
@@ -0,0 +1,256 @@
+                                break
+                        except DockerException:
+                            traceback.print_exc()
+                            time.sleep(1)
+
+                # Clean-up dependencies.
+                for dep in run_state.bundle['dependencies']:
+                    #self._run_manager.dependency_manager.remove_dependency(
+                    #    dep['parent_uuid'], dep['parent_path'], bundle_uuid)
+
+                    # Clean-up the symlinks we created.
+                    child_path = os.path.join(run_state.bundle_path, dep['child_path'])
+                    remove_path(child_path)
+
+                # Upload results
+                logger.debug('Uploading results for run with UUID %s', bundle_uuid)
+                def update_status(bytes_uploaded):
+                    run_status = 'Uploading results: %s done (archived size)' % size_str(bytes_uploaded)
+                    with self._run_manager.lock:
+                        self._run_manager.uploading[bundle_uuid]['run_status'] = run_status
+
+                self._run_manager.upload_bundle_contents(bundle_uuid, run_state.bundle_path, update_status)
+            except Exception:
+                traceback.print_exc()
+
+        bundle_uuid = run_state.bundle['uuid']
+        if bundle_uuid not in self._run_manager.uploading:
+            self._run_manager.uploading[bundle_uuid] = {
+                'thread': threading.Thread(target=upload_results, args=[]),
+                'run_status': ''
+            }
+            self._run_manager.uploading[bundle_uuid]['thread'].start()
+
+        if self._run_manager.uploading[bundle_uuid]['thread'].is_alive():
+            return run_state._replace(run_status=self._run_manager.uploading[bundle_uuid]['run_status'])
+        else: # thread finished
+            del self._run_manager.uploading[bundle_uuid]
+            return run_state._replace(stage=LocalRunStage.FINALIZING, container_id=None)
+
+    def _transition_from_FINALIZING(self, run_state):
+        def finalize():
+            try:
+                logger.debug('Finalizing run with UUID %s', bundle_uuid)
+                failure_message = run_state.info.get('failure_message', None)
+                exitcode = run_state.info.get('exitcode', None)
+                if failure_message is None and run_state.is_killed:
+                    failure_message = self._run_manager._get_kill_message()
+                finalize_message = {
+                    'exitcode': exitcode,
+                    'failure_message': failure_message,
+                }
+                self._run_manager.finalize_bundle(bundle_uuid, finalize_message)
+            except Exception:
+                traceback.print_exc()
+
+        bundle_uuid = run_state.bundle['uuid']
+        if bundle_uuid not in self._run_manager.finalizing:
+            self._run_manager.finalizing[bundle_uuid] = {
+                'thread': threading.Thread(target=finalize, args=[]),
```

Comment:
ditto about the "thread+return/err message" thingy

Reply:
+1